src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Implementation/ImplementationEndpoint.cs

```diff
@@ -0,0 +1,72 @@
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Implementation
+{
+    internal class ImplementationEndpoint : AbstractRazorDelegatingEndpoint<ImplementationParamsBridge, SumType<Location[]?, VSInternalReferenceItem[]?>>, IImplementationEndpoint
+    {
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public ImplementationEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
+            ILoggerFactory loggerFactory)
+            : base(documentContextFactory, languageServerFeatureOptions, documentMappingService, languageServer, loggerFactory.CreateLogger<ImplementationEndpoint>())
+        {
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
+
+        protected override string CustomMessageTarget => RazorLanguageServerCustomMessageTargets.RazorImplementationEndpointName;
+
+        public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
+        {
+            const string ServerCapability = "implementationProvider";
+            var option = new ImplementationOptions();
+
+            return new RegistrationExtensionResult(ServerCapability, option);
+        }
+
+        protected override IDelegatedParams CreateDelegatedParams(ImplementationParamsBridge request, DocumentContext documentContext, Projection projection, CancellationToken cancellationToken)
+            => new DelegatedPositionParams(
+                    documentContext.Identifier,
+                    projection.Position,
+                    projection.LanguageKind);
+
+        protected async override Task<SumType<Location[]?, VSInternalReferenceItem[]?>> HandleDelegatedResponseAsync(SumType<Location[]?, VSInternalReferenceItem[]?> delegatedResponse, DocumentContext documentContext, CancellationToken cancellationToken)
+        {
+            // Not using .TryGetXXX because this does the null check for us too
+            if (delegatedResponse.Value is Location[] locations)
+            {
+                foreach (var loc in locations)
+                {
+                    (loc.Uri, loc.Range) = await _documentMappingService.MapFromProjectedDocumentRangeAsync(loc.Uri, loc.Range, cancellationToken).ConfigureAwait(false);
+                }
+
+                return locations;
+            }
+            else if (delegatedResponse.Value is VSInternalReferenceItem[] referenceItems)
+            {
+                foreach (var item in referenceItems)
+                {
+                    (item.Location.Uri, item.Location.Range) = await _documentMappingService.MapFromProjectedDocumentRangeAsync(item.Location.Uri, item.Location.Range, cancellationToken).ConfigureAwait(false);
```

Comment:
The existing HtmlCSharp handler for GoToImpl calls a helper method to change the text of the results, if they're from generated code, but try as I might I couldn't get this to fail without it. I suspect the call was mainly there for the remapping behaviour, not the rewriting. The helper method is in `FindAllReferencesHelper`, and that functionality will definitely be needed for that endpoint when/if it moves, so we can always follow up if a user notices something.